java/src/main/java/org/whispersystems/signalservice/internal/websocket/WebSocketConnection.java

```diff
@@ -295,11 +304,21 @@ private long elapsedTime(long startTime) {
     private AtomicBoolean stop = new AtomicBoolean(false);
 
     public void run() {
+      final long timeoutMillis = TimeUnit.SECONDS.toMillis(KEEPALIVE_TIMEOUT_SECONDS);
+
       while (!stop.get()) {
-        try {
-          Thread.sleep(TimeUnit.SECONDS.toMillis(KEEPALIVE_TIMEOUT_SECONDS));
+        final long startTime = System.currentTimeMillis();
+
+        while (elapsedTime(startTime) < timeoutMillis) {
+          try {
+            sleepTimer.sleep(timeoutMillis - elapsedTime(startTime));
+          } catch (Throwable e) {
+            Log.w(TAG, e);
+          }
+        }
```

Comment:
Please move this while loop to where I originally asked for it. The goal here has always been for there to be zero logical changes to the existing GCM-supported code path. Making any logical changes to that code path at all makes it much more difficult to review.